### PR TITLE
Fix #378 - Hash IP addresses and User Agent information

### LINKED ISSUE #378: Hash IP addresses stored in database ?
IP Addresses can be PII (Personally Identifiable Information) and storing the cleartext value in the database is of no use to PHPAuth. Currently PHPAuth stores cleartext IP Addresses in the following places:

* `phpauth_attempts.ip`
* `phpauth_sessions.ip`

Instead we could store a hash of the IP address + site wide salt. This would mean that in case the database contents were stolen, the attacker would not be able to retrieve the original IP addresses. (This also requires storing the site wide salt outside of the database, aka in an `.env` file, which could possibly be auto-generated if not found on first run)

The same could also be done for User Agents, currently stored in cleartext in the `phpauth_sessions.agent` column.

What do you think ?

### PR DESCRIPTION
**Pull request by liamjack**
Hashes PII such as user's IP addresses and User Agent information.

This requires changing the structure of two tables:
* `phpauth_attempts`
* `phpauth_sessions`

A script (`upgrade.sql`) is provided, to delete all existing attempts and sessions from the database and update the table definitions.

This update also removes the `id` column from the `phpauth_attempts` table.

A new requirement is added: `vlucas/phpdotenv`, so that a `.env` file can be used to store a site-wide secret, that cannot be stored in the database. A new "Installation" section has been added to `README.md` to guide users.

### CODE REVIEW

**Comment by Conver** [comment]
Please do not approve and merge this yet!!! I'm coming home tomorrow and need to check a few things to make this compatible for everyone and not just as the current PR is.

**Comment by KarelWintersky** [comment]
Adding dotenv depencency for **ONE** configuration value?

Are you **crazy**???

I'll reject this.

**Comment by KarelWintersky** [comment]
Please, give **obvious** names to values. 

Correct name for `PHPAUTH_SECRET` is, really, `PHPAUTH_HASHING_SALT` 

Please, STOP making useless quasy-security changes.

File: Auth.php [hunk 1]
```diff
@@ -1518,6 +1521,29 @@ protected function checkReCaptcha($captcha_response)
...
+    private function getIpHash() {
+        return md5($this->getIp() . getenv("PHPAUTH_SECRET"), TRUE);
+    }
+
+    /**
+     * Returns the client's user agent
+     * @return string
+     */
+    private function getUserAgent() {
+        return isset($_SERVER['HTTP_USER_AGENT']) ? $_SERVER['HTTP_USER_AGENT'] : '';
```
Comment (KarelWintersky) [inline Auth.php]: Here you can wrote:

`return $_SERVER['HTTP_USER_AGENT'] ?? ''`; 

And we can now remove PHP5.6 depencency

**Comment by liamjack** [comment]
> Adding dotenv depencency for **ONE** configuration value?
> 
> Are you **crazy**???
> 
> I'll reject this.

Where do you propose storing the value then ? In the database ? That way if the database is stolen the attacker has the hashes and the salt required to create a rainbow table. Dotenv can be used in further commits for migrating more configuration variables away from the database, which is currently an inefficient storage mechanism.

> Please, give **obvious** names to values.
> 
> Correct name for `PHPAUTH_SECRET` is, really, `PHPAUTH_HASHING_SALT`
> 
> Please, STOP making useless quasy-security changes.

My bad, I'll rename to `PHPAUTH_HASHING_SALT`, but it still remains a value that needs to remain secret. What do you mean by "Please, STOP making useless quasy-security changes." ?

File: Auth.php (hunk 1 above)
Comment (liamjack) [reply Auth.php]: Great I'll fix this and get rid of PHP5.6 from composer, thanks.

**Comment by KarelWintersky** [comment]
> Dotenv can be used in further commits for migrating more configuration variables away 
> from the database, which is currently an inefficient storage mechanism.

if attacker recieved access to database - he recieves access to `.env` file too!  

> which is currently an inefficient storage mechanism.

Really?
Why don't use `config.json` or `config.ini` ? Or `config.php` with `return []` ?

**Comment by KarelWintersky** [comment]
> every single request to retrieve the application's configuration variables is inefficient yes.

Not for an ANY single request.
PHPAuth loaded once at application init and used via global AuthConfig class or any singletone wrapper. 

Loading configuration from file is already inefficient. 

Anyway, there is no any difference between loading from file or DB due caching. Better way is storing configuration in Redis. 

> with .env, making it easier to have different configuration files for different environments
> (production / development / staging) and switching between them.

All this differences can be implemented via `config.local.json`, `config.stable.json` etc. 

One problem: `.env` does not allow multi-dimensional configuration with sections.

**Comment by liamjack** [comment]
> Not for an ANY single request.
> PHPAuth loaded once at application init and used via global AuthConfig class or any singletone wrapper.

Singletons don't work like that in PHP: http://blog.gordon-oheim.biz/2011-01-17-Why-Singletons-have-no-use-in-PHP/

> Loading configuration from file is already inefficient.
>
> Anyway, there is no any difference between loading from file or DB due caching.

Do you have any numbers to back up these claims ?

> One problem: .env does not allow multi-dimensional configuration with sections.

Doesn't sound like much of a problem for PHPAuth's use case, does the database allow multi-dimensional configuration with sections for example ?

**Comment by KarelWintersky** [comment]
> Singletons don't work like that in PHP

Singleton is not a bad practice or good practice. Singleton is a tool for solve a problems.

> Do you have any numbers to back up these claims ?

Only logic :)

> does the database allow multi-dimensional configuration with sections for example 

in one of my projects I uses two database connections - one for connect to main database, second - for connection to Sphinx.

```
DB::init('main', $config_main); 
DB::init('sphinx', $config_sphinx);
...
DB::getConnection('main')->query('....');
DB::getConnection('sphinx')->query('...');
```
it is a different connections

**Comment by liamjack** [comment]
> Singleton is not a bad practice or good practice. Singleton is a tool for solve a problems.

I didn't say that singletons were good or bad. What I am trying to say is that upon every single request, a new Config object is instantiated, because each request has it's own memory. As said in the blog post I linked, PHP is a Shared-Nothing architecture. And so, each request = an SQL query to fetch PHPAuth's configuration variables from the database.

> Only logic :)

Well logically I would say that file based configuration would be more efficient. Databases can be local *or* remote, and can have more or less resources.

Instead of opening two database connections, you could just add a prefix in front of the variable name in the configuration file, or in the database.

**Comment by KarelWintersky** [comment]
> PHP is a Shared-Nothing architecture.

https://github.com/KarelWintersky/Arris

$c1 = DB::getConnection();
$c2 = DB::getConnection();
dump($c1 === $c2); // true

> And so, each request = an SQL query to fetch PHPAuth's configuration variables from the database.

Each request to application, not an each auth request. Each request to application loads a lot libs via composer, loads files, makes request to DB... etc.

Anyway, storing configuration on DB is not handy, but storing configuration in `.env` file is not handy too :(

I my projects I uses:

- `config.php` : https://github.com/KarelWintersky/LIBDb/blob/master/www/config/_config.example.php 
- `*.ini` : https://github.com/KarelWintersky/rpgclubs_xussr/blob/master/index.php 
- `*.ini` too : https://github.com/KarelWintersky/kwLME.LiveMap/blob/master/engine/__required.php (last example is not a good example, anyway)

> Instead of opening two database connections, you could just add a prefix in front 
> of the variable name in the configuration file, or in the database.

any example?

**Comment by liamjack** [comment]
> $c1 = DB::getConnection();
> $c2 = DB::getConnection();
> dump($c1 === $c2); // true

Yes that's a singleton and appears to be working correctly, but that's not what I'm talking about.
I'm talking about HTTP requests to the application. Every HTTP request = a query to fetch configuration variables = inefficient: 500 requests = 500 database queries to get the same information.

> any example?

```
PHPAUTH_VARIABLE_1="foo"
PHPAUTH_VARIABLE_2="bar"

SPHINX_VARIABLE_1="foo"
SPHINX_VARIABLE_2="bar"
```

**Comment by KarelWintersky** [comment]
> variables = inefficient: 500 requests = 500 database queries to get the same information.

1st fetching data from database, 
499 - loading from database cache.

Don't think that database makers are idiots. Good caching settings boost efficient.

with file:
1st loading from disk
499 - loading from memory cache :)

**Comment by KarelWintersky** [comment]
> PHPAUTH_VARIABLE_1="foo"
> PHPAUTH_VARIABLE_2="bar"
> SPHINX_VARIABLE_1="foo"
> SPHINX_VARIABLE_2="bar"

And how I must load this configuration? 
INI file is better:
```
[db:main]
host = 'localhost'
port = 3306
user = 'karel'

[db:sphinx]
host = '127.0.0.1'
port = 6033    ; port for dockerized sphinx application
user = 'sphinx'
```

**Comment by liamjack** [comment]
Well in that case, feel free to make the changes yourself.

**Comment by louis123562** [comment]
XML > INI

Thats my opinion...

I dont know how big your configurations oder how croppy your servers are, but **i wont change anything on the current config logic**.

**Comment by Conver** [comment]
So where do we stand on this one?

**Comment by KarelWintersky** [comment]
ATM, I think, we must use configuration storage like this:

https://packagist.org/packages/adbario/php-dot-notation

Source of configuration can be ANY (ini, xml, yaml, etc), but all must return array of options. 

any config -> loader (outside of PHPAuth) that will return [] -> PHPAuthConfig storage -> PHPAuth. 

We must specify configuration format: 
```
[
'cookie.domain'  => NULL,
'cookie.forget'  => '+30 minutes',
'site.page.activation' => 'activate',
'site.email' => '...'
'site.key' => '...' 
'site.url' => '... ',
'smtp._' =>  0,
'smtp.debug' => 0
'smtp.host' => '...'
'table.attempts' => 'phpauth_attempts',
'table.sessions' => 'phpauth_sessions',
'verify.email.max_length' => 100,
'verify.email.use_banlist' => true,
'translation.source' => 'php'
'recaptcha.enabled' => false,
...
]
```

and:
```
<?php

namespace PHPAuth;

use Adbar\Dot;

/**
 * Class AppConfig
 * @package Arris
 */
class PHPAuthConfig implements PHPAuthConfigInterface
{
    protected static $instance = null;

    /**
     *
     * @param Dot $instance
     * @return Dot
     */
    public static function init(Dot &$instance)
    {
        self::$instance = $instance;
        return $instance;
    }

    /**
     *
     * @return Dot
     */
    public static function get(): Dot
    {
        return self::$instance;
    }

    /**
     *
     * @param Dot $instance
     */
    public static function set(Dot &$instance)
    {
        self::$instance = $instance;
    }
}
```

Now, we can use it:
```
$config = new \Adbar\Dot(  load config here );
PHPAuthConfig::init( $config );
```
and later:
```
PHPAuthConfig::get('table.attempts')
```
or

function foobar()
{
  $CONFIG = PHPAuthConfig::get(); // so, we use global singletone ^_^
  $key_session = $CONFIG['auth.key_session'];
  $key_cookie  = $CONFIG['auth.key_cookie'];
}

**Comment by Conver** [comment]
status?

**Comment by KarelWintersky** [comment]
Cooking V2 Version :)

**Comment by KarelWintersky** [comment]
Will be available something like:
```
try {
    $dbh = new PDO("mysql:host=localhost;dbname=phpauth", "wombat", "wombatsql");
    
    $config = new Config($dbh);

    $config->loadDB($dbh, 'phpauth_config_admins');

    // or    
    $config->loadArray(['bbb' => 15])->loadINI('$/test_v2.ini')->locale('ru_RU')->getConfig();

    // or    
    $config->loadJSON('test_v2.json')->getConfig();

    // or    
    $config->loadYAML('test_v2.yaml');
    
    $config->foobar = 15;

    $config->dictionary['account_not_found'] = 'XXXX';
    var_dump($config->dictionary['account_not_found']);

    $p = new Auth($dbh, $config );
    .... 

} catch (\PHPAuth\PHPAuthException $e) {
    var_dump($e->getMessage());
}


```

**Comment by Conver** [comment]
@KarelWintersky - Let me know if you want help on v2 :+1:

**Pull request closed without merging**